pkg/controller/master/image/vm_image_controller.go

```diff
@@ -57,6 +57,26 @@ func (h *vmImageHandler) OnChanged(_ string, image *harvesterv1.VirtualMachineIm
 			return h.images.Update(toUpdate)
 		}
 
+		// sync virtualSize (handles the case for existing images that were
+		// imported before this field was added, because adding the field to
+		// the CRD triggers vmImageHandler.OnChanged)
+		if image.Status.VirtualSize == 0 {
+			bi, err := util.GetBackingImage(h.backingImageCache, image)
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					logrus.WithError(err).WithFields(logrus.Fields{
+						"namespace": image.Namespace,
+						"name":      image.Name,
+					}).Error("failed to get backing image for vmimage")
+					return image, err
+				}
+				return h.handleRetry(image)
+			}
+			toUpdate := image.DeepCopy()
+			toUpdate.Status.VirtualSize = bi.Status.VirtualSize
+			return h.images.Update(toUpdate)
+		}
```

Comment:
IMHO, for an imported image, we should not make it go through `hanldeRetry()`. It will delete related bi and sc, and recreate all resources again. For the case bi is missing, maybe we can just log the error, and update the image condition. WDYT?

Reply:
Oh, good catch. Yeah, I agree, given the backing image and storage class really _should_ be there already for an imported image, deleting those things in this case seems like a bad idea.

I definitely think we should log the error, but what do you think the condition should be updated to?

Reply:
https://github.com/harvester/harvester/blob/9c0a3a17f42831b590a92f3743ac027a66a25b8d/pkg/apis/harvesterhci.io/v1beta1/image.go#L9-L13

just my 2 cents, maybe a new condition like `BackingImageMiss`?

Reply:
Alright, will see what I can do tomorrow!

Reply:
OK, the latest change will add a BackingImageMissing condition to the vmimage on error here, with a message set to the error, which means you can see this if something goes wrong:

```
# kubectl describe vmimages/image-kqcnt
[...]
Status:
  Conditions:
    Last Update Time:  2024-07-04T02:08:48Z
    Reason:            Imported
    Status:            True
    Type:              Imported
    Last Update Time:  2024-07-04T02:08:32Z
    Reason:            Initialized
    Status:            True
    Type:              Initialized
    Last Update Time:  2024-07-04T02:35:07Z
    Message:           Failed to get backing image: backingimages.longhorn.io "default-image-kqcnt" not found
    Status:            True
    Type:              BackingImageMissing
  Failed:              0
  Progress:            100
  Size:                260465664
  Storage Class Name:  longhorn-image-kqcnt
  Virtual Size:        833617920
Events:                <none>
```

As I mentioned in the comment in the code, the image isn't actually marked failed or anything in case this error is transient. I really think it's extremely unlikely to hit this, so I haven't added any code to automatically clear that condition if the image later comes good somehow (I don't know how such an event would be triggered anyway).

I tested this by:
- deleting the backing image
- upgrading the harvester container
- verifying that the error was logged and that the condition was set on the image

I then repaired the backing image by:
- using the longhorn GUI to upload a backing image with the same name as the missing image
- edited the image to tweak its description (which triggers the OnChange code, which in turn triggers the virtual size change, which worked this time because the backing image was present)
- manually deleted the BackingImageMissing condition with `kubectl edit`